src/server/frontend_wayland/xdg_activation_v1.cpp

```diff
@@ -200,68 +188,97 @@ mf::XdgActivationV1::XdgActivationV1(
     std::shared_ptr<ms::SessionCoordinator> const& session_coordinator,
     std::shared_ptr<MainLoop> const& main_loop,
     std::shared_ptr<ObserverRegistrar<input::KeyboardObserver>> const& keyboard_observer_registrar,
-    Executor& wayland_executor)
+    Executor& wayland_executor,
+    std::shared_ptr<msh::TokenAuthority> token_authority)
     : Global(display, Version<1>()),
       shell{shell},
       session_coordinator{session_coordinator},
       main_loop{main_loop},
       keyboard_observer_registrar{keyboard_observer_registrar},
-      keyboard_observer{std::make_shared<KeyboardObserver>(this)}
+      keyboard_observer{std::make_shared<KeyboardObserver>(this)},
+      session_listener{std::make_shared<SessionListener>(this)},
+      token_authority{token_authority}
 {
     keyboard_observer_registrar->register_interest(keyboard_observer, wayland_executor);
+    session_coordinator->add_listener(session_listener);
 }
 
 mf::XdgActivationV1::~XdgActivationV1()
 {
     keyboard_observer_registrar->unregister_interest(*keyboard_observer);
+    session_coordinator->remove_listener(session_listener);
 }
 
-std::shared_ptr<mf::XdgActivationTokenData> const& mf::XdgActivationV1::create_token(std::shared_ptr<ms::Session> const& session)
+std::shared_ptr<mf::XdgActivationTokenData> mf::XdgActivationV1::get_token_data(Token const& token)
 {
-    auto generated = generate_token();
-    auto token = std::make_shared<XdgActivationTokenData>(generated, main_loop->create_alarm([this, generated]()
+    std::lock_guard guard(pending_tokens_mutex);
+
+    auto iter = std::ranges::find_if(
+        pending_tokens,
+        [&](auto const& pending_token)
+        {
+            return pending_token->token == token;
+        });
+
+    if (iter != pending_tokens.end())
+        return *iter;
+
+    return nullptr;
+}
+
+std::shared_ptr<mf::XdgActivationTokenData> const mf::XdgActivationV1::create_token(std::shared_ptr<ms::Session> const& session)
+{
+    if (!session_listener->is_focused_session(session))
     {
-        try_consume_token(generated);
-    }), session);
+        mir::log_warning(
+            "Client from a different session trying to create a token. Session name: %s, PID: %d. Returning bogus "
```

Comment:
```suggestion
            "Client from a unfocussed session trying to create a token. Session name: %s, PID: %d. Returning bogus "
```

Also, I don't think frontend_wayland populates session names. So that may be uninformative.

From the discussion about client behaviour, this might spam the logs.